File: app/src/main/java/com/pr0gramm/app/services/RulesService.java
package com.pr0gramm.app.services;

import android.content.res.Resources;
import android.widget.TextView;

import com.google.common.base.Strings;
import com.pr0gramm.app.R;
import com.pr0gramm.app.util.AndroidUtility;
import com.pr0gramm.app.util.BackgroundScheduler;
import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;
import com.trello.rxlifecycle.RxLifecycle;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.util.async.Async;

/**
 */
@Singleton
public class RulesService {
    private static final String DEFAULT_RULES_TEXT = "<li>nsfw/nsfl Bilder müssen vor dem Upload entsprechend markiert werden!</li> <li>Keine suggestiven Bilder oder Gore von/mit Minderjährigen/Babys/Föten.</li> <li>Keine Tierpornos.</li> <li>Kein stumpfer Rassismus, kein rechtes Gedankengut, keine Nazi-Nostalgie.</li> <li>Keine Werbung; keine Affiliate Links in den Bildern; kein Spam.</li> <li>Keine Informationen oder Bilder von Privatpersonen; Keine Klarnamen in den Uploads, Tags oder Kommentaren.</li> <li>Ein Mindestmaß an Bildqualität wird erwartet. Bildmaterial mit starken Kompressionsartefakten, übermäßig großen Watermarks oder unsinnig beschnittene/skalierte Bilder werden gelöscht.</li> <li>Keine Bilder mit ähnlichem Inhalt in Reihe. Zugehöriger Content kann in den Kommentaren verlinkt werden.</li> <li> Kommentare wie <em>“Tag deinen Scheiß”</em> und ähnliches gehören nicht in die Tags. Genaueres im FAQ: <a href=\"#faq:tags\">Was gehört in die Tags?</a> </li> <li>Downvote-Spam, Vote-Manipulation und Tag-Vandalismus werden nicht geduldet.</li> <li>Pro Benutzer ist nur ein Account erlaubt. Indizien für Multiaccounts sind gegenseitige Upvotes oder Spamaktionen.</li> <li>Keine Warez, gestohlene Logins zu Pay Sites o.ä.</li> <li>Überzogene oder häufige Beleidigungen anderen Benutzern und insbesondere gegenüber Moderatoren wird mit einer Sperrung bestraft.</li>";

    private final Observable<String> rules;

    @Inject
    public RulesService(OkHttpClient okHttpClient) {
        this.rules = Async.fromCallable(() -> {
                    long cacheSlayer = System.currentTimeMillis() / (24 * 3600);
                    String url = "https://pr0gramm.com/media/pr0gramm.min.js?app" + cacheSlayer;
                    Request request = new Request.Builder().url(url).build();
                    Response response = okHttpClient.newCall(request).execute();
                    return response.isSuccessful() ? response.body().string() : null;
                },

                /* run on background scheduler */
                BackgroundScheduler.instance())

                /* skip errors */
                .filter(responseText -> !Strings.isNullOrEmpty(responseText))
                .onErrorResumeNext(Observable.empty())

                /* extract from javascript */
                .flatMap(this::extractRulesFromJavascript)

                /* start with the default */
                .startWith(DEFAULT_RULES_TEXT)

                /* Cache the last one */
                .replay(1).autoConnect();
    }

    public void displayInto(TextView targetView) {
        this.rules
                .subscribeOn(BackgroundScheduler.instance())
                .observeOn(AndroidSchedulers.mainThread())
                .compose(RxLifecycle.bindView(targetView))

                /* display the result */
                .subscribe(text -> displayInto(targetView, text));
    }

    private Observable<? extends String> extractRulesFromJavascript(String response) {
        Matcher matcher = Pattern.compile("Rules=[^']+'([^']+)").matcher(response);
        return matcher.find() ? Observable.just(matcher.group(1)) : Observable.empty();
    }

    private void displayInto(TextView rulesView, String rules) {
        int idx = 0;
        List<String> list = new ArrayList<>();
        Matcher matcher = Pattern.compile("<li>(.+?)</li>").matcher(rules);
        while (matcher.find()) {
            String rule = matcher.group(1).replaceAll("<[^>]+>", "").trim();
            list.add((idx + 1) + ". " + rule);
        }

        Resources resources = rulesView.getContext().getResources();
        int offset = resources.getDimensionPixelSize(R.dimen.bullet_list_leading_margin);
        rulesView.setText(AndroidUtility.makeBulletList(offset, list));
    }
}

File: app/src/main/java/com/pr0gramm/app/services/RulesService.kt
package com.pr0gramm.app.services

import android.widget.TextView
import com.pr0gramm.app.R
import com.pr0gramm.app.util.AndroidUtility
import com.pr0gramm.app.util.ignoreAllExceptions
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runInterruptible
import okhttp3.OkHttpClient
import okhttp3.Request

/**
 */
class RulesService(private val okHttpClient: OkHttpClient) {
    private var cachedText: String = defaultRulesText

    suspend fun displayInto(targetView: TextView) {
        displayInto(targetView, cachedText)

        // now try to fetch the updated text
        ignoreAllExceptions {
            val text = runInterruptible(Dispatchers.IO) {
                val url = "https://pr0gramm.com/media/templates/rules.html"
                val request = Request.Builder().url(url).build()
                val response = okHttpClient.newCall(request).execute()
                if (response.isSuccessful) response.body?.string() else null
            }

            if (text != null) {
                // display the new text again on success
                displayInto(targetView, text)
            }
        }
    }

    private fun displayInto(rulesView: TextView, rules: String) {
        val list = "<pre>(.+?)</pre>".toRegex(RegexOption.DOT_MATCHES_ALL).findAll(rules).mapIndexed { _, match ->
            var id = "";
            var rule = match.groupValues[1].replace("<b>(.+?)</b>".toRegex(RegexOption.IGNORE_CASE)) { id = it.groupValues[1]; ""}
            rule = rule.replace("<[^>]+>".toRegex(), "").trim { it <= ' ' }

            "#$id  $rule"
        }

        val resources = rulesView.context.resources
        val offset = resources.getDimensionPixelSize(R.dimen.bullet_list_leading_margin)
        rulesView.text = AndroidUtility.makeBulletList(offset, list.toList())
    }
}

private const val defaultRulesText = """
        <pre><b>1</b> nsfp/nsfw/nsfl Inhalte müssen vor dem Upload entsprechend markiert werden.</pre>
		<pre><b>2</b> <i>Minderjährige:</i></pre>
		<pre><b>2.1</b> Keine suggestive oder nackte Darstellung von Minderjährigen.</pre>
		<pre><b>2.2</b> Keine rohe Gewalt an Minderjährigen.</pre>
		<pre><b>3</b> <i>Tiere:</i></pre>
		<pre><b>3.1</b> Keine Tierquälerei.</pre>
		<pre><b>3.2</b> Keine Zoophilie oder Fetischvideos mit Tieren.</pre>
		<pre><b>4</b> <i>Rassismus und Hetze:</i></pre>
		<pre><b>4.1</b> Kein stumpfer Rassismus, kein rechtes Gedankengut.</pre>
		<pre><b>4.2</b> Keine Hetze, egal ob politisch, rassistisch oder religiös motiviert.</pre>
		<pre><b>4.3</b> Keine Aufrufe zu Gewalt.</pre>
		<pre><b>4.4</b> Keine Nazi-Nostalgie/Nazi-Nostalgia</pre>
		<pre><b>4.5</b> Kein Agenda Pushing oder Verbreiten von Propaganda.</pre>
		<pre><b>5</b> Keine Werbung</pre>
		<pre><b>6</b> Keine Informationen oder Bilder/Videos von Privatpersonen.</pre>
		<pre><b>7</b> <i>Contentqualität:</i></pre>
		<pre><b>7.1</b> Ein Mindestmaß an Bild/Videoqualität wird erwartet.</pre>
		<pre><b>7.2</b> Reposts gilt es zu vermeiden.</pre>
		<pre><b>7.3</b> Keine Müllposts/Privatmüll.</pre>
		<pre><b>7.4</b> Keine langweiligen Raids.</pre>
		<pre><b>8</b> Keine Bilder/Videos mit ähnlichem Inhalt kurz hintereinander posten.</pre>
		<pre><b>9</b> <i>Vandalismus:</i></pre>
		<pre><b>9.1</b> Kein Tag-Vandalismus.</pre>
		<pre><b>9.2</b> Kein Spam/Kommentarvandalismus.</pre>
		<pre><b>9.3</b> Kein Downvote-Spam.</pre>
		<pre><b>9.4</b> Keine Vote-Manipulation.</pre>
		<pre><b>9.5</b> Kein Downvoten sinnvoller Tags.</pre>
		<pre><b>10</b> Keine frühen Spoiler.</pre>
		<pre><b>11</b> Pro Benutzer ist nur ein Account erlaubt.</pre>
		<pre><b>12</b> Keine Warez, Links zu illegalen Angeboten, gestohlene Logins zu Pay Sites o.Ä.</pre>
		<pre><b>13</b> Keine übermäßigen Beleidigungen oder Hetzen gegen andere Benutzer, die Community oder die Moderation.</pre>
		<pre><b>14</b> Keine “Screamer” oder sonstige Ton-Videos mit der Absicht, Benutzer zu erschrecken oder zu trollen.</pre>
		<pre><b>15</b> <i>Musikuploads:</i></pre>
		<pre><b>15.1</b> Keine reinen Musikuploads.</pre>
		<pre><b>15.2</b> Keine Musikvideos.</pre>
		<pre><b>16</b> <i>Störungen der Moderation:</i></pre>
		<pre><b>16.1</b> Kein unnötiges Markieren von Moderatoren oder Nerven von Community-Helfern.</pre>
		<pre><b>16.2</b> Kein Missbrauch der Melden-Funktion.</pre>
    """